File: fortran/repl.f90
program repl
  implicit none
  character(len=256) :: buf
  integer :: length
  character(:), allocatable :: history_file
  character(*), parameter :: path = '/.mal_history'
  character(*), parameter :: space_chars = " 	" // new_line('')
  call get_environment_variable('HOME', buf, length)
  allocate(character(length + len(path)) :: history_file)
  history_file = buf(:length) // path
  call main('> ')
contains
  subroutine load_history()
    use, intrinsic :: iso_fortran_env
    use iso_readline
    character(len=256) :: line
    integer :: unit, size, iostat
    open(newunit=unit, file=history_file, action='read', &
         iostat=iostat, form='formatted', access='stream')
    if (iostat > 0) return
    do
       read(unit, '(a)', advance='no', iostat=iostat, size=size) line
       if (iostat == IOSTAT_END) exit
       call add_history(line(:size))
    end do
    close(unit)
  end subroutine load_history

  logical function is_blank(string)
    character(*), intent(in) :: string
    is_blank = verify(string, space_chars) == 0
  end function is_blank

  subroutine append_line(path, line)
    character(*), intent(in) :: path, line
    integer :: unit
    open(newunit=unit, file=path, action='write', &
         position='append', form='formatted', access='stream')
    write(unit, '(a)') line
    close(unit)
  end subroutine append_line

  subroutine add_to_history(line)
    use iso_readline
    character(*), intent(in) :: line
    if (.not. is_blank(line)) then
       call add_history(line)
       call append_line(history_file, line)
    end if
  end subroutine add_to_history

  subroutine main(prompt)
    use iso_readline
    character(*), intent(in) :: prompt
    character(len=256) :: line
    integer :: length
    call load_history()
    do while (.true.)
       call readline(prompt, line, length)
       if (length < 0) exit
       if (.not. is_blank(line(:length))) then
          write(*, '(g0)') line(:length)
          call add_to_history(line(:length))
       end if
    end do
    write(*, '(g0)') ''
  end subroutine main
end program repl

File: c/repl.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <readline/readline.h>
#include <readline/history.h>

#define MAX_STRING_LENGTH 256
char history_file[MAX_STRING_LENGTH];

void chomp(char *buf) {
    int len = (buf == NULL) ? 0 : strlen(buf);
    if (len > 0 && buf[len-1] == '\n') {
        buf[len-1] = '\0';
    }
}

void load_history() {
    char line[MAX_STRING_LENGTH];
    FILE *f = fopen(history_file, "r");
    if (f != NULL) {
        while(1) {
            char *ret = fgets(line, MAX_STRING_LENGTH, f);
            if (feof(f) && ret == NULL) {
                break;
            } else if (ret == NULL) {
                line[MAX_STRING_LENGTH-1] = '\0';
            }

            chomp(line);
            add_history(line);
        }
        fclose(f);
    }
}

int is_blank(char *line) {
    if (line == NULL) {
        return 1;
    }

    int len = strlen(line);
    int blank_len = strspn(line, " \t\n");
    return (len == blank_len);
}

void append_line(char *path, char *line) {
    FILE* f = fopen(path, "a");
    if (f != NULL) {
        fprintf(f, "%s\n", line);
        fclose(f);
    }
}

void add_to_history(char *line) {
    if (!is_blank(line)) {
        add_history(line);
        append_line(history_file, line);
    }
}

void repl(char *prompt) {
    char *line;
    load_history();
    while ((line = readline(prompt)) != NULL) {
        if (!is_blank(line)) {
            printf("%s\n", line);
            add_to_history(line);
        }
        free(line);
    }
    printf("\n");
}

int main() {
    char *home = getenv("HOME");
    char *path = "/.mal_history";
    assert(home != NULL);
    snprintf(history_file, MAX_STRING_LENGTH, "%s%s", home, path);
    repl("> ");
    return 0;
}
